### DISCUSSION TITLE
Prebuilt wheels provided via 3rd party repository

### DISCUSSION THREAD

**Discussion by MiroPsota** [facebookresearch/pytorch3d#1752] [Show and tell]
## Prebuilt wheels for PyTorch packages with custom ops

I've created a repository that can build PyTorch wheels with custom ops through the GitHub Actions pipeline and publish them using GitHub Releases, with build attestation. Check it out at https://github.com/MiroPsota/torch_packages_builder.

Since there are various ways how to use it, please refer to the repository README for more information.

If you prefer own build or can't trust a 3rd party repository, feel free to fork it and build any package/version/commit ID you desire yourself.

- **Pip cache might not work**: pip relies on http cache, and GitHub generates on-the-fly redirections for release links, so they are probably not playing nicely together. If you use uv, you can set custom cache policy.

### PyTorch3D specific info:

Install using pip:

```sh
pip install --extra-index-url https://miropsota.github.io/torch_packages_builder pytorch3d==<version>+<OPTIONAL_commit_hash>pt<PyTorch_version><compute_platform>
```

Where `<compute_platform>` is, as in PyTorch, one of `cpu`, `cu<CUDA_short_version>` (e.g. `cu130`, `cu128`, `cu118`), or `rocm<ROCM_version>` (not supported right now).

For example, the newest pytorch3d commit (as of writing) `d9839a9`, PyTorch 2.13.0 with CUDA 13.2:

```sh
pip install --extra-index-url https://miropsota.github.io/torch_packages_builder pytorch3d==0.7.9+d9839a9pt2.13.0cu132
```

Look at releases section if there are any other combinations, I will probably build occasionally with new pytorch releases and versions/commits.

These wheels are built with PyTorch versions `2.3.0` to `2.13.0` (check my GitHub releases for older versions) and their respective compute platforms and supported operating systems. Please note an exception for the ROCm platform. The build is done using the available public GitHub runners, so they might not work on older OSes.

Although the wheels have been successfully built, I don't test them to work correctly.

If you've installed PyTorch with pip, there's no need to have CUDA installed on your system, as the PyTorch wheels for pip bundle CUDA.

**Comment by yueool** [comment]
谢谢你，自己编译好几次都失败了，还是用现成的吧，我好笨，仰望大佬

**Comment by randprint** [comment]
can you check the build logs for windows 2.2.2 cu118? I tried to install from your repo to fix the dll import problem i'm having with torch3d but it seems the dll that should be in Lib\site-packages\pytorch3d is missing, if i understand the response to this issue that I'm also having https://github.com/facebookresearch/pytorch3d/issues/1664

**Reply by MiroPsota** [reply]
If I understand it correctly, Just do as the [last post](https://github.com/facebookresearch/pytorch3d/issues/1664#issuecomment-1774837924) says.

**Reply by randprint** [reply]
tried that out of frustration, -e expects some file to be found on the url, but it isnt part of any pytorch3d install instructions I've seen anywhere anyway.

**Comment by Jacky-MYQ** [comment]
Wo! amazing!! I try it successfully! And my env is py3.11, cuda12.1, ubuntu22.04. I sincerely appreciate it that does me a favor!!

**Comment by zhanghua7099** [comment]
Thanks for your great work! This helped me a lot.

**Comment by shewangmu** [comment]
呜呜呜，好人一生平安

**Comment by FurkanGozukara** [comment]
so awesome gonna test now

pip install --extra-index-url https://miropsota.github.io/torch_packages_builder pytorch3d==0.7.8+pt2.5.1cu124

**Reply by mkat2** [reply]
This saved me!  Thanks! for pytorch3d !  even did with cu128 and pt2.7.1

**Comment by brianlan** [comment]
Works for me, thanks. @MiroPsota MiroPsota

**Comment by linden713** [comment]
Thanks A LOT! You really have saved our life!

**Comment by axbycc-mark** [comment]
Great! This deserves support from the official team plus a mention in the installation guide.

**Comment by lly007** [comment]
Great work！
I suggest pytorch3d officially put this in the readme.

**Comment by UpendraArun** [comment]
Very helpful! This should go on the official install readme.

**Comment by rajk97** [comment]
Thank you very much!

**Comment by eddiesmithxyz** [comment]
You are sent from the heavens

**Comment by dolienky** [comment]
You are a GOD!

**Comment by Yicheng-Cao** [comment]
OMG, thank you so much for your contribution. It took me about one day to build the wheel, but it kept failing. You don't know how much this has helped me.

**Comment by tuqizhao** [comment]
@MiroPsota Can you provide the torch 2.6.0  cuda 12.6  python 3.12 on  windows version， Although  is not stable now. Tks

**Reply by MiroPsota** [reply]
Unfortunately, it's not possible at the moment, even with the `main` branch.

**Comment by fcbfcb1998** [comment]
Awesome

**Comment by guansy123** [comment]
Thank you very much!!!

**Comment by HaoxiangYou** [comment]
You are the hero!!!!!

**Comment by FurkanGozukara** [comment]
anyone managed to compile for CUDA 12.8 and Torch 2.7? pytorch3d

**Reply by mkat2** [reply]
yes!

**Reply by mkat2** [reply]
pip list | grep pytorch3d
pytorch3d                               0.7.8+pt2.7.1cu128

**Comment by ZZFyigeshuaige** [comment]
thanks!

**Comment by WY-2022** [comment]
Cool!

**Comment by wangtngx** [comment]
thanks

**Comment by fleurs03** [comment]
rescued me from a time crunch

**Comment by Taokt** [comment]
Thank you so much this saved me lot of time!!

**Comment by Rookienovice** [comment]
好人一生平安！

**Comment by hairuoliu1** [comment]
Amzaing!

**Comment by xinyu-pu** [comment]
Thank you so much!

**Comment by cekay-S** [comment]
Thank you very much MiroPsota! It is working for me now and saved me some trouble.

**Comment by CMCC011** [comment]
Thank you a lot for your work!!! It finally works on my machine. My env is python3.11, cuda12.4, windows10, and I have torch2.6.0+cu124 and VS2022. I kept meeting some "cl.exe", "-std=c++17", and C1083 errors, but now it works successfully with pip install --extra-index-url https://miropsota.github.io/torch_packages_builder pytorch3d==0.7.8+5043d15pt2.6.0cu124

Thank you again for sharing this repository.

**Comment by rohit7044** [comment]
The messiah !!

**Comment by nb824798** [comment]
thank you very much！

**Comment by MrShoenel** [comment]
Great work! I am using Cuda 12.8 with Python 3.14 and Pytorch 2.10 and it works nicely, thank you.

**Comment by NinZeige** [comment]
For `uv` users, one can configure `pyproject.toml` like this:

```toml
dependencies = [
    "pytorch3d==0.7.9+pt2.8.0cu126",
]

[tool.uv.sources]
pytorch3d = { index = "miropsota" }


[[tool.uv.index]]
name = "miropsota"
url = "https://miropsota.github.io/torch_packages_builder"
```
